apps/emqx_message_transformation/src/emqx_message_transformation.erl

```diff
@@ -0,0 +1,820 @@
+
+-type transformation_name() :: binary().
+%% TODO
+-type transformation() :: #{atom() => term()}.
+%% TODO
+-type variform() :: any().
+-type operation() :: #{key := [binary(), ...], value := variform()}.
+-type qos() :: 0..2.
+-type rendered_value() :: qos() | boolean() | binary().
+
+-type eval_context() :: #{
+    payload := _,
+    qos := _,
+    retain := _,
+    topic := _,
+    user_property := _,
+    dirty := #{
+        payload => true,
+        qos => true,
+        retain => true,
+        topic => true,
+        user_property => true
+    }
+}.
+
+%%------------------------------------------------------------------------------
+%% API
+%%------------------------------------------------------------------------------
+
+-spec add_handler() -> ok.
+add_handler() ->
+    ok = emqx_config_handler:add_handler([?CONF_ROOT], ?MODULE),
+    ok = emqx_config_handler:add_handler(?TRANSFORMATIONS_CONF_PATH, ?MODULE),
+    ok.
+
+-spec remove_handler() -> ok.
+remove_handler() ->
+    ok = emqx_config_handler:remove_handler(?TRANSFORMATIONS_CONF_PATH),
+    ok = emqx_config_handler:remove_handler([?CONF_ROOT]),
+    ok.
+
+load() ->
+    Transformations = emqx:get_config(?TRANSFORMATIONS_CONF_PATH, []),
+    lists:foreach(
+        fun({Pos, Transformation}) ->
+            ok = emqx_message_transformation_registry:insert(Pos, Transformation)
+        end,
+        lists:enumerate(Transformations)
+    ).
+
+unload() ->
+    Transformations = emqx:get_config(?TRANSFORMATIONS_CONF_PATH, []),
+    lists:foreach(
+        fun(Transformation) ->
+            ok = emqx_message_transformation_registry:delete(Transformation)
+        end,
+        Transformations
+    ).
+
+-spec list() -> [transformation()].
+list() ->
+    emqx:get_config(?TRANSFORMATIONS_CONF_PATH, []).
+
+-spec reorder([transformation_name()]) ->
+    {ok, _} | {error, _}.
+reorder(Order) ->
+    emqx_conf:update(
+        ?TRANSFORMATIONS_CONF_PATH,
+        {reorder, Order},
+        #{override_to => cluster}
+    ).
+
+-spec lookup(transformation_name()) -> {ok, transformation()} | {error, not_found}.
+lookup(Name) ->
+    Transformations = emqx:get_config(?TRANSFORMATIONS_CONF_PATH, []),
+    do_lookup(Name, Transformations).
+
+-spec insert(transformation()) ->
+    {ok, _} | {error, _}.
+insert(Transformation) ->
+    emqx_conf:update(
+        ?TRANSFORMATIONS_CONF_PATH,
+        {append, Transformation},
+        #{override_to => cluster}
+    ).
+
+-spec update(transformation()) ->
+    {ok, _} | {error, _}.
+update(Transformation) ->
+    emqx_conf:update(
+        ?TRANSFORMATIONS_CONF_PATH,
+        {update, Transformation},
+        #{override_to => cluster}
+    ).
+
+-spec delete(transformation_name()) ->
+    {ok, _} | {error, _}.
+delete(Name) ->
+    emqx_conf:update(
+        ?TRANSFORMATIONS_CONF_PATH,
+        {delete, Name},
+        #{override_to => cluster}
+    ).
+
+%%------------------------------------------------------------------------------
```

Comment:
Nit: it seems that the above funs (together with pre/post config update hooks) are not coupled with the rest of the module (which is large). They seem to use only registry and local helpers.
Maybe move to a separate module like `emqx_message_transformation_config` in some other apps?

Reply:
:heavy_check_mark: